alibi/utils/distributed.py

```diff
@@ -217,3 +232,523 @@ def _chunk(values, chunksize):
+
+                class DistributedOpts(TypedDict):
+                    n_cpus: Optional[int]
+                    batch_size: Optional[int]
+            The dictionary may contain two additional keys:
+                
+                - ``'actor_cpu_frac'`` (float, <= 1.0, >0.0): This is used to create more than one process on one \\
+                CPU/GPU. This may not speed up CPU intensive tasks but it is worth experimenting with when few physical \\ 
+                cores are available. In particular, this is highly useful when the user wants to share a GPU for \\
+                multiple tasks. See the ``ray`` documentation `here_` for details.
+                
+                .. _here:
+                   https://docs.ray.io/en/stable/resources.html#fractional-resources
+                
+                - ``'algorithm'``: this is specified internally by the caller. It is used in order to register callbacks \\
+                for results postprocessing or target functions for the parallel pool. These should be implemented in the \\
+                global scope. If not specified, its value will be ``'default'``, meaning that the results will not be \\
+                post-processed and that the default target function will be used. 
+        explainer_type
+            Explainer class.
+        explainer_init_args
+            Positional argument to explainer constructor.
+        explainer_init_kwargs
+            Keyword arguments to explainer constructor.
+        return_generator
+            If `True` a generator that returns the results in the order the computation finishes is returned when
+            `get_explanation` is called. Otherwise, the order of the results is the same as the order of the minibatches.
+            
+        Notes
+        -----
+        When `return_generator=True`, the caller has to take elements from the generator (e.g., by calling `next`) in 
+        order to start computing the results (because the `ray` pool is implemented as a generator).
+        """  # noqa W605
+
+        # TODO: FINISH DOCSTRING
+        # TODO: TBD: @JANIS - DO WE HAVE TO DO ANY SETUP FOR THIS COMMAND TO WORK?
+        if not RAY_INSTALLED:
+            raise ModuleNotFoundError("Module requires ray to be installed. pip install alibi[ray] ")
+
+        self.n_processes = distributed_opts['n_cpus']
+        self.batch_size = distributed_opts['batch_size']
+        self.return_generator = return_generator
+        algorithm = distributed_opts.get('algorithm', 'default')
+        if 'algorithm' == 'default':
+            logger.warning(
+                "No algorithm specified in distributed option, default target function and postprocessing will be "
+                "selected."
+            )
+        self.target_fcn = default_target_fcn
+        # TODO: TBD: @JANIS - THE POSTPROCESSING COULD BE HANDLED BY CALLER? OR THE CALLER COULD PASS THE FUNCTION?
```

Comment:
Differences:

- target function: used by ray to schedule a remote explainer call (see `submit` method in `ray.util.ActorPool`

- postprocessing function: used to re-assemble the results computed in parallel into a single object (think e.g., concatenation of numpy arrays in the simplest case). It needs to be custom because each explainer returns something different. We can have a generic one that assumes the input array is a list numpy arrays and concatenates them, and register different functions for algorithms with more complex output types (e.g., kernel shap). 

[NB: On second thoughts I am thinking that `singledispatch` might be another way to implement this]